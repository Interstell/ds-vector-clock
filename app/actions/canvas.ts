import { Stage } from 'react-konva';
import {
  eventType,
  processType,
  CHANGE_CANVAS_SIZE,
  MESSAGE_SENDING_STARTED,
  SET_STAGE_REF,
  MESSAGE_SENDING_ABORTED,
  PROCESS_CREATE,
  EVENT_CREATE,
  MESSAGE_SENDING_FINISHED,
  RESET_STATE
} from '../reducers/canvas';

export function resetState() {
  return {
    type: RESET_STATE
  };
}

export function changeCanvasSize(width: number, height: number) {
  return {
    type: CHANGE_CANVAS_SIZE,
    payload: {
      width,
      height
    }
  };
}

export function setStageRef(stageRef: Stage) {
  return {
    type: SET_STAGE_REF,
    payload: {
      stageRef
    }
  };
}

export function startSendingMessage(messageSendingEvent: eventType) {
  return {
    type: MESSAGE_SENDING_STARTED,
    payload: { messageSendingEvent }
  };
}

export function abortSendingMessage() {
  return {
    type: MESSAGE_SENDING_ABORTED
  };
}

export function finishSendingMessage(eventTo: eventType) {
  return {
    type: MESSAGE_SENDING_FINISHED,
    payload: {
      eventTo
    }
  };
}

export function createNewProcess() {
  return {
    type: PROCESS_CREATE
  };
}

// todo delete process

export function createNewEvent(process: processType, x: number) {
  return {
    type: EVENT_CREATE,
    payload: {
      x,
      process
    }
  };
}

// todo delete event
