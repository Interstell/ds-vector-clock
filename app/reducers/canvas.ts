import { AnyAction } from 'redux';
import { Stage } from 'react-konva';

export const prefix = '@map/';
export const CHANGE_CANVAS_SIZE = `${prefix}CHANGE_CANVAS_SIZE`;
export const SET_STAGE_REF = `${prefix}SET_STAGE_REF`;

export const MESSAGE_SENDING_STARTED = `${prefix}MESSAGE_SENDING_STARTED`;
export const MESSAGE_SENDING_ABORTED = `${prefix}MESSAGE_SENDING_ABORTED`;
export const MESSAGE_SENDING_FINISHED = `${prefix}MESSAGE_SENDING_FINISHED`;

export const PROCESS_CREATE = `${prefix}PROCESS_CREATE`;

export const EVENT_CREATE = `${prefix}EVENT_CREATE`;

export const RESET_STATE = `${prefix}RESET_STATE`;

export type eventType = {
  id: string;
  processId: string;
  x: number;
};

export type processType = {
  id: string;
  events?: eventType[];
};

export type messageType = {
  id: string;
  eventFrom: eventType;
  eventTo: eventType;
};

export type canvasStateType = {
  canvasWidth: number;
  canvasHeight: number;

  processes: processType[];
  messages: messageType[];

  stageRef?: Stage;

  isMessageSendingState: boolean;
  messageSendingEvent?: eventType;

  processNameCounter: number;
  eventNameCounter: number;
};

function createDefaultState(): canvasStateType {
  const defaultState: canvasStateType = {
    canvasWidth: 900,
    canvasHeight: 600,

    processes: [
      {
        id: 'p1',
        events: [
          {
            id: 'A',
            processId: 'p1',
            x: 120
          },
          {
            id: 'B',
            processId: 'p1',
            x: 200
          }
        ]
      },
      {
        id: 'p2',
        events: [
          {
            id: 'C',
            processId: 'p2',
            x: 150
          }
        ]
      },
      {
        id: 'p3',
        events: [
          { id: 'D', processId: 'p3', x: 100 },
          { id: 'E', processId: 'p3', x: 250 }
        ]
      }
    ],
    messages: [],

    isMessageSendingState: false,

    processNameCounter: 4,
    eventNameCounter: 5
  };

  defaultState.messages.push({
    id: 'A->C',
    // @ts-ignore
    eventFrom: defaultState.processes[0].events[0],
    // @ts-ignore
    eventTo: defaultState.processes[1].events[0]
  });

  return defaultState;
}

export default function canvas(
  state = createDefaultState(),
  action: AnyAction
) {
  switch (action.type) {
    case RESET_STATE:
      return createDefaultState();
    case CHANGE_CANVAS_SIZE:
      return {
        ...state,
        canvasWidth: action.payload.width,
        canvasHeight: action.payload.height
      };
    case SET_STAGE_REF:
      return {
        ...state,
        stageRef: action.payload.stageRef
      };
    case MESSAGE_SENDING_STARTED:
      return {
        ...state,
        isMessageSendingState: true,
        messageSendingEvent: action.payload.messageSendingEvent
      };
    case MESSAGE_SENDING_ABORTED:
      return {
        ...state,
        isMessageSendingState: false,
        messageSendingEvent: null
      };
    case MESSAGE_SENDING_FINISHED:
      return {
        ...state,
        isMessageSendingState: false,
        messageSendingEvent: null,
        messages: [
          ...state.messages,
          {
            id: `${state.messageSendingEvent?.id}->${action.payload.eventTo.id}`,
            eventFrom: state.messageSendingEvent,
            eventTo: action.payload.eventTo
          }
        ]
      };
    case PROCESS_CREATE:
      return {
        ...state,
        processes: [
          ...state.processes,
          { id: `p${state.processNameCounter}`, events: [] }
        ],
        processNameCounter: state.processNameCounter + 1
      };
    case EVENT_CREATE: {
      const processIndex = state.processes.findIndex(
        p => p === action.payload.process
      );
      return {
        ...state,
        processes: [
          ...state.processes.slice(0, processIndex),
          {
            ...action.payload.process,
            events: [
              ...action.payload.process.events,
              {
                id: String.fromCharCode(
                  state.eventNameCounter <= 25
                    ? 65 + state.eventNameCounter
                    : 97 + (state.eventNameCounter % 26)
                ),
                processId: action.payload.process.id,
                x: action.payload.x
              }
            ]
          },
          ...state.processes.slice(processIndex + 1)
        ],
        eventNameCounter: state.eventNameCounter + 1
      };
    }

    default:
      return state;
  }
}
