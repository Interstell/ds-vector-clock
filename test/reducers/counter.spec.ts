import canvas from '../../app/reducers/canvas';
import {
  INCREMENT_COUNTER,
  DECREMENT_COUNTER
} from '../../app/actions/canvas';

describe('reducers', () => {
  describe('counter', () => {
    it('should handle initial state', () => {
      expect(canvas(undefined, {})).toMatchSnapshot();
    });

    it('should handle INCREMENT_COUNTER', () => {
      expect(canvas(1, { type: INCREMENT_COUNTER })).toMatchSnapshot();
    });

    it('should handle DECREMENT_COUNTER', () => {
      expect(canvas(1, { type: DECREMENT_COUNTER })).toMatchSnapshot();
    });

    it('should handle unknown action type', () => {
      expect(canvas(1, { type: 'unknown' })).toMatchSnapshot();
    });
  });
});
