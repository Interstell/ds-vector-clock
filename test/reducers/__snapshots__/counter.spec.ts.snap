// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers canvas should handle DECREMENT_COUNTER 1`] = `0`;

exports[`reducers canvas should handle INCREMENT_COUNTER 1`] = `2`;

exports[`reducers canvas should handle initial state 1`] = `0`;

exports[`reducers canvas should handle unknown action type 1`] = `1`;
